From 2fab0912b3e62fd3c51a65029d832394e8e75121 Mon Sep 17 00:00:00 2001
From: Jagpal Singh Gill <paligill@gmail.com>
Date: Tue, 31 Mar 2026 19:45:25 -0700
Subject: [PATCH] configuration: add AnalogValve D-Bus interface

Add D-Bus interface definitions for analog valve configuration used in
liquid cooling systems. This includes:

- AnalogValve: main interface with properties for valve name, type,
  max flow rate, setpoint voltage range, tolerance, and open threshold.
- SetPointDAC: sub-interface for DAC device configuration with I2C
  bus/address, channel index, offset, and scale factor.
- FeedbackADC: sub-interface for ADC device configuration used for
  valve position feedback with channel index and scale factor.

Also add the corresponding generated meson build files.

Reference:
https://gerrit.openbmc.org/c/openbmc/entity-manager/+/88760
https://gerrit.openbmc.org/c/openbmc/entity-manager/+/88761

Change-Id: I9f27ce1f238c05faf05d075e34525b2b0d847674
Signed-off-by: Jagpal Singh Gill <paligill@gmail.com>
---
 .../AnalogValve/FeedbackADC/meson.build       | 40 +++++++++
 .../AnalogValve/SetPointDAC/meson.build       | 40 +++++++++
 .../Configuration/AnalogValve/meson.build     | 90 +++++++++++++++++++
 .../openbmc_project/Configuration/meson.build | 25 ++++++
 .../Configuration/AnalogValve.interface.yaml  | 42 +++++++++
 .../AnalogValve/FeedbackADC.interface.yaml    | 29 ++++++
 .../AnalogValve/SetPointDAC.interface.yaml    | 47 ++++++++++
 7 files changed, 313 insertions(+)
 create mode 100644 gen/xyz/openbmc_project/Configuration/AnalogValve/FeedbackADC/meson.build
 create mode 100644 gen/xyz/openbmc_project/Configuration/AnalogValve/SetPointDAC/meson.build
 create mode 100644 gen/xyz/openbmc_project/Configuration/AnalogValve/meson.build
 create mode 100644 yaml/xyz/openbmc_project/Configuration/AnalogValve.interface.yaml
 create mode 100644 yaml/xyz/openbmc_project/Configuration/AnalogValve/FeedbackADC.interface.yaml
 create mode 100644 yaml/xyz/openbmc_project/Configuration/AnalogValve/SetPointDAC.interface.yaml

diff --git a/gen/xyz/openbmc_project/Configuration/AnalogValve/FeedbackADC/meson.build b/gen/xyz/openbmc_project/Configuration/AnalogValve/FeedbackADC/meson.build
new file mode 100644
index 0000000..5032044
--- /dev/null
+++ b/gen/xyz/openbmc_project/Configuration/AnalogValve/FeedbackADC/meson.build
@@ -0,0 +1,40 @@
+# Generated file; do not modify.
+
+sdbusplus_current_path = 'xyz/openbmc_project/Configuration/AnalogValve/FeedbackADC'
+
+generated_sources += custom_target(
+    'xyz/openbmc_project/Configuration/AnalogValve/FeedbackADC__cpp'.underscorify(),
+    input: [
+        '../../../../../../yaml/xyz/openbmc_project/Configuration/AnalogValve/FeedbackADC.interface.yaml',
+    ],
+    output: [
+        'common.hpp',
+        'server.hpp',
+        'server.cpp',
+        'aserver.hpp',
+        'client.hpp',
+    ],
+    depend_files: sdbusplusplus_depfiles,
+    command: [
+        sdbuspp_gen_meson_prog,
+        '--command',
+        'cpp',
+        '--output',
+        meson.current_build_dir(),
+        '--tool',
+        sdbusplusplus_prog,
+        '--directory',
+        meson.current_source_dir() / '../../../../../../yaml',
+        'xyz/openbmc_project/Configuration/AnalogValve/FeedbackADC',
+    ],
+    install: should_generate_cpp,
+    install_dir: [
+        get_option('includedir') / sdbusplus_current_path,
+        get_option('includedir') / sdbusplus_current_path,
+        false,
+        get_option('includedir') / sdbusplus_current_path,
+        get_option('includedir') / sdbusplus_current_path,
+    ],
+    build_by_default: should_generate_cpp,
+)
+
diff --git a/gen/xyz/openbmc_project/Configuration/AnalogValve/SetPointDAC/meson.build b/gen/xyz/openbmc_project/Configuration/AnalogValve/SetPointDAC/meson.build
new file mode 100644
index 0000000..794c7c1
--- /dev/null
+++ b/gen/xyz/openbmc_project/Configuration/AnalogValve/SetPointDAC/meson.build
@@ -0,0 +1,40 @@
+# Generated file; do not modify.
+
+sdbusplus_current_path = 'xyz/openbmc_project/Configuration/AnalogValve/SetPointDAC'
+
+generated_sources += custom_target(
+    'xyz/openbmc_project/Configuration/AnalogValve/SetPointDAC__cpp'.underscorify(),
+    input: [
+        '../../../../../../yaml/xyz/openbmc_project/Configuration/AnalogValve/SetPointDAC.interface.yaml',
+    ],
+    output: [
+        'common.hpp',
+        'server.hpp',
+        'server.cpp',
+        'aserver.hpp',
+        'client.hpp',
+    ],
+    depend_files: sdbusplusplus_depfiles,
+    command: [
+        sdbuspp_gen_meson_prog,
+        '--command',
+        'cpp',
+        '--output',
+        meson.current_build_dir(),
+        '--tool',
+        sdbusplusplus_prog,
+        '--directory',
+        meson.current_source_dir() / '../../../../../../yaml',
+        'xyz/openbmc_project/Configuration/AnalogValve/SetPointDAC',
+    ],
+    install: should_generate_cpp,
+    install_dir: [
+        get_option('includedir') / sdbusplus_current_path,
+        get_option('includedir') / sdbusplus_current_path,
+        false,
+        get_option('includedir') / sdbusplus_current_path,
+        get_option('includedir') / sdbusplus_current_path,
+    ],
+    build_by_default: should_generate_cpp,
+)
+
diff --git a/gen/xyz/openbmc_project/Configuration/AnalogValve/meson.build b/gen/xyz/openbmc_project/Configuration/AnalogValve/meson.build
new file mode 100644
index 0000000..36620ac
--- /dev/null
+++ b/gen/xyz/openbmc_project/Configuration/AnalogValve/meson.build
@@ -0,0 +1,90 @@
+# Generated file; do not modify.
+subdir('FeedbackADC')
+subdir('SetPointDAC')
+
+sdbusplus_current_path = 'xyz/openbmc_project/Configuration/AnalogValve'
+
+generated_sources += custom_target(
+    'xyz/openbmc_project/Configuration/AnalogValve__cpp'.underscorify(),
+    input: [
+        '../../../../../yaml/xyz/openbmc_project/Configuration/AnalogValve.interface.yaml',
+    ],
+    output: [
+        'common.hpp',
+        'server.hpp',
+        'server.cpp',
+        'aserver.hpp',
+        'client.hpp',
+    ],
+    depend_files: sdbusplusplus_depfiles,
+    command: [
+        sdbuspp_gen_meson_prog,
+        '--command',
+        'cpp',
+        '--output',
+        meson.current_build_dir(),
+        '--tool',
+        sdbusplusplus_prog,
+        '--directory',
+        meson.current_source_dir() / '../../../../../yaml',
+        'xyz/openbmc_project/Configuration/AnalogValve',
+    ],
+    install: should_generate_cpp,
+    install_dir: [
+        get_option('includedir') / sdbusplus_current_path,
+        get_option('includedir') / sdbusplus_current_path,
+        false,
+        get_option('includedir') / sdbusplus_current_path,
+        get_option('includedir') / sdbusplus_current_path,
+    ],
+    build_by_default: should_generate_cpp,
+)
+
+generated_markdown += custom_target(
+    'xyz/openbmc_project/Configuration/AnalogValve/FeedbackADC__markdown'.underscorify(),
+    input: [
+        '../../../../../yaml/xyz/openbmc_project/Configuration/AnalogValve/FeedbackADC.interface.yaml',
+    ],
+    output: ['FeedbackADC.md'],
+    depend_files: sdbusplusplus_depfiles,
+    command: [
+        sdbuspp_gen_meson_prog,
+        '--command',
+        'markdown',
+        '--output',
+        meson.current_build_dir(),
+        '--tool',
+        sdbusplusplus_prog,
+        '--directory',
+        meson.current_source_dir() / '../../../../../yaml',
+        'xyz/openbmc_project/Configuration/AnalogValve/FeedbackADC',
+    ],
+    install: should_generate_markdown,
+    install_dir: [inst_markdown_dir / sdbusplus_current_path],
+    build_by_default: should_generate_markdown,
+)
+
+generated_markdown += custom_target(
+    'xyz/openbmc_project/Configuration/AnalogValve/SetPointDAC__markdown'.underscorify(),
+    input: [
+        '../../../../../yaml/xyz/openbmc_project/Configuration/AnalogValve/SetPointDAC.interface.yaml',
+    ],
+    output: ['SetPointDAC.md'],
+    depend_files: sdbusplusplus_depfiles,
+    command: [
+        sdbuspp_gen_meson_prog,
+        '--command',
+        'markdown',
+        '--output',
+        meson.current_build_dir(),
+        '--tool',
+        sdbusplusplus_prog,
+        '--directory',
+        meson.current_source_dir() / '../../../../../yaml',
+        'xyz/openbmc_project/Configuration/AnalogValve/SetPointDAC',
+    ],
+    install: should_generate_markdown,
+    install_dir: [inst_markdown_dir / sdbusplus_current_path],
+    build_by_default: should_generate_markdown,
+)
+
diff --git a/gen/xyz/openbmc_project/Configuration/meson.build b/gen/xyz/openbmc_project/Configuration/meson.build
index 409ba98..a72314f 100644
--- a/gen/xyz/openbmc_project/Configuration/meson.build
+++ b/gen/xyz/openbmc_project/Configuration/meson.build
@@ -1,4 +1,5 @@
 # Generated file; do not modify.
+subdir('AnalogValve')
 subdir('GPIODeviceDetect')
 subdir('GPIOLeakDetector')
 subdir('GPIOValve')
@@ -8,6 +9,30 @@ subdir('USBPort')
 
 sdbusplus_current_path = 'xyz/openbmc_project/Configuration'
 
+generated_markdown += custom_target(
+    'xyz/openbmc_project/Configuration/AnalogValve__markdown'.underscorify(),
+    input: [
+        '../../../../yaml/xyz/openbmc_project/Configuration/AnalogValve.interface.yaml',
+    ],
+    output: ['AnalogValve.md'],
+    depend_files: sdbusplusplus_depfiles,
+    command: [
+        sdbuspp_gen_meson_prog,
+        '--command',
+        'markdown',
+        '--output',
+        meson.current_build_dir(),
+        '--tool',
+        sdbusplusplus_prog,
+        '--directory',
+        meson.current_source_dir() / '../../../../yaml',
+        'xyz/openbmc_project/Configuration/AnalogValve',
+    ],
+    install: should_generate_markdown,
+    install_dir: [inst_markdown_dir / sdbusplus_current_path],
+    build_by_default: should_generate_markdown,
+)
+
 generated_markdown += custom_target(
     'xyz/openbmc_project/Configuration/GPIODeviceDetect__markdown'.underscorify(),
     input: [
diff --git a/yaml/xyz/openbmc_project/Configuration/AnalogValve.interface.yaml b/yaml/xyz/openbmc_project/Configuration/AnalogValve.interface.yaml
new file mode 100644
index 0000000..7c2c89c
--- /dev/null
+++ b/yaml/xyz/openbmc_project/Configuration/AnalogValve.interface.yaml
@@ -0,0 +1,42 @@
+description: >
+    Implement to provide entity manager configuration for an analog valve used
+    in liquid cooling.
+
+properties:
+    - name: Type
+      type: string
+      description: >
+          The type of configuration object.
+      flags:
+          - readonly
+    - name: Name
+      type: string
+      description: >
+          The name of the analog valve.
+      flags:
+          - readonly
+    - name: MaxFlowRate
+      type: double
+      description: >
+          The maximum flow rate supported by the analog valve.
+      flags:
+          - readonly
+    - name: SetPointVoltageRange
+      type: array[double]
+      description: >
+          The voltage range [min, max] for the valve setpoint.
+      flags:
+          - readonly
+    - name: Tolerance
+      type: double
+      description: >
+          The acceptable voltage deviation between setpoint and feedback before
+          the valve is considered out of position.
+      flags:
+          - readonly
+    - name: OpenThreshold
+      type: double
+      description: >
+          The threshold voltage above which the valve is considered open.
+      flags:
+          - readonly
diff --git a/yaml/xyz/openbmc_project/Configuration/AnalogValve/FeedbackADC.interface.yaml b/yaml/xyz/openbmc_project/Configuration/AnalogValve/FeedbackADC.interface.yaml
new file mode 100644
index 0000000..afb3888
--- /dev/null
+++ b/yaml/xyz/openbmc_project/Configuration/AnalogValve/FeedbackADC.interface.yaml
@@ -0,0 +1,29 @@
+description: >
+    Implement to provide entity manager configuration for an analog to digital
+    converter (ADC) device used for valve position feedback.
+
+properties:
+    - name: Name
+      type: string
+      description: >
+          The name of the ADC device.
+      flags:
+          - readonly
+    - name: Type
+      type: string
+      description: >
+          The type of configuration object.
+      flags:
+          - readonly
+    - name: Index
+      type: uint64
+      description: >
+          The channel index of the ADC device.
+      flags:
+          - readonly
+    - name: ScaleFactor
+      type: double
+      description: >
+          The scale factor for ADC raw value conversion.
+      flags:
+          - readonly
diff --git a/yaml/xyz/openbmc_project/Configuration/AnalogValve/SetPointDAC.interface.yaml b/yaml/xyz/openbmc_project/Configuration/AnalogValve/SetPointDAC.interface.yaml
new file mode 100644
index 0000000..bf65391
--- /dev/null
+++ b/yaml/xyz/openbmc_project/Configuration/AnalogValve/SetPointDAC.interface.yaml
@@ -0,0 +1,47 @@
+description: >
+    Implement to provide entity manager configuration for a digital to analog
+    converter (DAC) device.
+
+properties:
+    - name: Type
+      type: string
+      description: >
+          The type of configuration object.
+      flags:
+          - readonly
+    - name: Name
+      type: string
+      description: >
+          The name of the DAC device.
+      flags:
+          - readonly
+    - name: Bus
+      type: uint64
+      description: >
+          The I2C bus number.
+      flags:
+          - readonly
+    - name: Address
+      type: uint64
+      description: >
+          The I2C address of the DAC device.
+      flags:
+          - readonly
+    - name: Index
+      type: uint64
+      description: >
+          The channel index of the DAC device.
+      flags:
+          - readonly
+    - name: Offset
+      type: double
+      description: >
+          The voltage offset for DAC raw value conversion.
+      flags:
+          - readonly
+    - name: ScaleFactor
+      type: double
+      description: >
+          The scale factor for DAC raw value conversion.
+      flags:
+          - readonly
-- 
2.34.1

